File: ruby/spec/04_trees_and_graphs/07_build_order.rb
# n - count of projects
# Time = O(n^2) Mem = O(n)
def build_order_a(projects, deps)
  isolated = {}

  nodes = projects.each_with_object({}) do |x, res|
    res[x] = DirectedGraph::Node.new(x)
    isolated[x] = true
  end

  deps.each do |(from, to)|
    nodes[from].pointers << nodes[to]
    isolated[from] = false
    isolated[to] = false
  end

  nodes.each do |name, node|
    if res = has_route_to_all_nodes?(node, nodes, isolated)
      if no_cyces?(node)
        return res
      end
    end
  end

  nil
end

def has_route_to_all_nodes?(node, nodes, isolated)
  covered = isolated.select { |_, i| i == true }.to_h
  q = [node]
  path = covered.keys.sort

  while !q.empty? do
    curr = q.shift
    covered[curr.value] = true
    path << curr.value

    curr.pointers.each do |node|
      if !covered[node.value]
        q << node
        covered[node.value] = true
      end
    end
  end

  if covered.size == nodes.size
    return path
  end
end

def no_cyces?(node)
  covered_edges = Hash.new { |hash, key| hash[key] = {} } # to -> from
  q = [node]

  while !q.empty? do
    curr = q.shift

    curr.pointers.each do |node|
      if covered_edges[node.value][curr.value]
        return false
      elsif covered_edges[node.value].empty?
        q << node
      end

      covered_edges[node.value][curr.value] = true
    end
  end

  true
end

RSpec.describe 'build_order' do
  %i(a).each do |implementation|
    describe "#{implementation} case" do
      subject { send("build_order_#{implementation}", projects, deps) }

      let(:projects) { %i(a b c d e f) }

      context 'solvable deps' do
        # [:a, :d] -> means :d depends on :a
        let(:deps) { [[:a, :d], [:f, :b], [:b, :d], [:f, :a], [:d, :c]] }

        it { is_expected.to eq [:e, :f, :b, :a, :d, :c] }
      end

      context 'unsolvable deps' do
        let(:deps) { [[:a, :d], [:f, :b], [:b, :d], [:f, :a], [:d, :c], [:c, :f]] }

        it { is_expected.to eq nil }
      end
    end
  end
end

File: go/04_trees_and_graphs/07_build_order.go
package trees_and_graphs

import (
	. "github.com/holyketzer/ctci_v6/lib"
	"sort"
)

// n - count of projects
// Time = O(n^2) Mem = O(n)
func BuildOrderA(projects []int, deps [][]int) []int {
	isolated := make(map[int]bool)
	nodes := make(map[int]*DGNode)

	for _, x := range projects {
		nodes[x] = &DGNode{Value: x}
		isolated[x] = true
	}

	for _, pair := range deps {
		from := pair[0]
		to := pair[1]

		nodes[from].Pointers = append(nodes[from].Pointers, nodes[to])
		isolated[from] = false
		isolated[to] = false
	}

	for _, node := range nodes {
		res := HasRouteToAllNodes(node, nodes, isolated)

		if res != nil {
			if NoCyces(node) {
				return res
			}
		}
	}

	return nil
}

func HasRouteToAllNodes(node *DGNode, nodes map[int]*DGNode, isolated map[int]bool) []int {
	covered := make(map[int]bool)

	for name, i := range isolated {
		if i {
			covered[name] = true
		}
	}

	path := make([]int, 0, len(covered))

	for k, _ := range covered {
		path = append(path, k)
	}

	sort.Ints(path)

	q := []*DGNode{node}
	var curr *DGNode

	for len(q) > 0 {
		curr, q = q[0], q[1:]
		covered[curr.Value] = true
		path = append(path, curr.Value)

		for _, node := range curr.Pointers {
			if covered[node.Value] == false {
				q = append(q, node)
				covered[node.Value] = true
			}
		}
	}

	if len(covered) == len(nodes) {
		return path
	} else {
		return nil
	}
}

func NoCyces(node *DGNode) bool {
	// Hash.new { |hash, key| hash[key] = {} }
	coveredEdges := make(map[int](map[int]bool)) // to -> from
	q := []*DGNode{node}
	var curr *DGNode

	for len(q) > 0 {
		curr, q = q[0], q[1:]

		for _, node := range curr.Pointers {
			edge := coveredEdges[node.Value]
			edgePresent := false

			if edge != nil {
				edgePresent = edge[curr.Value]
			}

			if edgePresent {
				return false
			} else if coveredEdges[node.Value] == nil {
				q = append(q, node)
			}

			if edge == nil {
				coveredEdges[node.Value] = make(map[int]bool)
				coveredEdges[node.Value][curr.Value] = true
			}
		}
	}

	return true
}
